---
title: "pWGBSSimla: DMRcaller DMRs"
author: "Deepak Tanwar"
date: "<b>Created on:</b> 2020-02-14 <br> <b>Updated on:</b> `r Sys.Date()`"
output:
  rmdformats::readthedown:
    highlight: tango
    self_contained: true
    thumbnails: true
    lightbox: true
    gallery: false
    keep_md: no
    number_sections: no
    fig_width: 8
    fig_height: 8
    fig_caption: true
    df_print: paged
    code_folding: show
  fontsize: 12pt
  geometry: margin=1in
  documentclass: article
# bibliography: references.bib
link-citations: yes
---


Analysis following the [DMRcaller vignette](https://bioconductor.org/packages/release/bioc/vignettes/DMRcaller/inst/doc/DMRcaller.pdf).

# Packages required
```{r, warning=F, message=FALSE}
library(DT)
library(dplyr)
library(data.table)
library(DMRcaller)
```

# Method description

**DMRcaller offers three different methods of preprocessing data:**

1. **Neighbourhood method:** Each cytosine is considered independently, differentially methylated cytosines are called.

2. **Bins method:** The genome is split into tilling bins of given size, all reads in each bin are pooled and differentially methylated bins are called.

3. **Noise filter method:** A smoothing kernel is applied on the data before calling differentially methylated loci. Four kernels are available: uniform, triangular, Gaussian, and Epanechnicov.

For each cytosine (in case of `neighbourhood` and `noise filter` methods) or bin (in case of `bins` method) a statistical test (Fisherâ€™s exact or Score test) is performed. All statistically significant positions with at least given difference in methylation proportion and given number of reads per cytosine are called differentially methylated. Adjacent DMLs/bins are then merged.

Unlike the rest of the methods, DMRcaller allows to call DMRs also in non-CpG context, making it suitable e.g. for analysis of plant genomes.


# Input data

## Simulated data
```{r}
anno.sim <- read.table("./input/anno_sim_data.txt.gz", header = T, sep = "\t", stringsAsFactors = F)
files.sim <- list.files("./input", pattern = "sim_r", full.names = T)
files.sim <- files.sim[c(2:10, 1)]
read.sim <- GRangesList(lapply(files.sim, function(x) readBismark(x)))
names(read.sim) <- anno.sim$samples
mData.grp1.sim <- poolMethylationDatasets(read.sim[anno.sim$samples[anno.sim$Group == "Group1"]])
mData.grp2.sim <- poolMethylationDatasets(read.sim[anno.sim$samples[anno.sim$Group == "Group2"]])
```


# DMRs analysis
Two different statistical tests are available - Fisher's exact and Score test. Since according to the original paper both of them should perform equally well, we have decided to use computationally much less expensive Score test.

The p-values of the test are already corrected for multiple testing.

## Function to perform the analysis
```{r}
dmrcaller_analysis <- function(grp1, grp2, methods = "all", kernals = NULL, keyword) {

  # Methods
  m <- NULL
  if (methods == "all") {
    m <- c("noise_filter", "neighbourhood", "bins")
  } else {
    m <- methods
  }

  # Kernals
  k <- NULL
  if (methods == "noise_filter" | methods == "all") {
    k <- c("uniform", "triangular", "gaussian", "epanechnicov")
  } else {
    k <- kernels
  }
  # DMRs
  dmr.list.nf <- list()
  dmr.list.ot <- list()
  for (i in 1:length(m)) {
    if (m[i] == "noise_filter") {
      # run for all kernels
      for (j in 1:length(k)) {
        dmr.list.nf[[j]] <- computeDMRs(
          methylationData1 = grp1,
          methylationData2 = grp2,
          method = m[i],
          kernelFunction = k[j],
          test = "score",
          pValueThreshold = 0.9,
          minCytosinesCount = 10,
          minProportionDifference = 0.1,
          cores = detectCores() - 1
        )
        names(dmr.list.nf)[j] <- paste(m[i], k[j], sep = "_")
      }
    } else {
      dmr.list.ot[[i]] <- computeDMRs(
        methylationData1 = grp1,
        methylationData2 = grp2,
        method = m[i],
        test = "score",
        pValueThreshold = 0.9,
        minCytosinesCount = 10,
        minProportionDifference = 0.1,
        cores = detectCores() - 1
      )
      names(dmr.list.ot)[i] <- m[i]
    }
  }
  dmr.list <- c(dmr.list.nf, dmr.list.ot)
  dmr.list <- dmr.list[!is.na(names(dmr.list))]

  # iteratively merge DMRs after the analysis
  dmr.list.merge <- lapply(dmr.list, function(x) {
    mergeDMRsIteratively(
      DMRs = x,
      methylationData1 = grp1,
      methylationData2 = grp2,
      minGap = 200,
      respectSigns = TRUE,
      minProportionDifference = 0.1,
      pValueThreshold = 0.9,
      test = "score",
      cores = detectCores() - 1
    )
  })

  # Change colname to be compatible with the downstream analysis
  dmr.list.merge <- lapply(dmr.list.merge, function(x) {
    a <- data.frame(x, stringsAsFactors = F)
    colnames(a)[colnames(a) == "pValue"] <- "qval"
    return(GRanges(a))
  })

  # Save data

  for (i in 1:length(dmr.list.merge)) {
    tmp <- data.frame(dmr.list.merge[[i]])
    # colnames(tmp)[15] <- "qval"
    if (nrow(tmp) != 0) {
      n <- paste0("./output/DMRcaller_dmr_", names(dmr.list.merge)[i], "_", keyword, ".txt")
      write.table(tmp, file = n, quote = F, sep = "\t", row.names = F)
    }
  }

  return(dmr.list.merge)
}
```

## Simulated data

```{r}
dmr.ls.sim <- dmrcaller_analysis(
  grp1 = mData.grp1.sim, grp2 = mData.grp2.sim,
  methods = "all", kernals = "all",
  keyword = "sim_data"
)
```


# DMR tables {.tabset .tabset-pills}

```{r, echo=FALSE,include = FALSE}
# You need this code to conduct the magic dependences attaching...
DT::datatable(matrix())
```

## Simulated data {.tabset .tabset-dropdown}
```{r, results='asis'}
for (i in seq_along(dmr.ls.sim)) {
  name <- names(dmr.ls.sim)[i]
  cat("\n \n")
  cat(paste("###", name))
  cat("\n \n")
  print(htmltools::tagList(DT::datatable(data.frame(dmr.ls.sim[[i]]),
    rownames = F, width = "100%"
  )))
  cat("\n \n")
}
```


# Compress output
```{r}
system("pigz -11 -p 16 ./output/*.txt")
```


# SessionInfo

```{r}
devtools::session_info()
```
